code/parachain/frame/pablo/src/dual_asset_constant_product.rs

```diff
@@ -80,68 +84,233 @@ impl<T: Config> DualAssetConstantProduct<T> {
-		let first_asset_amount =
-			assets.get(first_asset).expect("Must exist as per previous validations");
-		let second_asset = assets_vec.get(1).expect("Must exist as per previous validations");
-		let second_asset_amount =
-			assets.get(second_asset).expect("Must exist as per previous validations");
+	) -> Result<T::Balance, DispatchError> {
+		let mut pool_assets = Self::get_pool_balances(&pool, &pool_account);
+
+		let assets_with_balances = assets
+			.iter()
+			.map(|asset_amount| {
+				if asset_amount.amount.is_zero() {
+					return Err(Error::<T>::InvalidAmount)
+				};
+
+				let balance =
+					pool_assets.remove(&asset_amount.asset_id).ok_or(Error::<T>::AssetNotFound)?;
+
+				Ok((asset_amount, balance))
+			})
+			.collect::<Result<Vec<_>, _>>()?;
 
 		let lp_total_issuance = T::Convert::convert(T::Assets::total_issuance(pool.lp_token));
-		let (quote_amount, amount_of_lp_token_to_mint) = compute_deposit_lp(
-			lp_total_issuance,
-			T::Convert::convert(*first_asset_amount),
-			T::Convert::convert(*second_asset_amount),
-			pool_assets.get(first_asset).expect("Must exist as per previous validations").1,
-			pool_assets.get(second_asset).expect("Must exist as per previous validations").1,
-		)?;
-		let quote_amount = T::Convert::convert(quote_amount);
+
+		let amount_of_lp_token_to_mint = match assets_with_balances[..] {
+			[(single, (single_weight, single_balance))] => {
+				if lp_total_issuance.is_zero() {
+					return Err(Error::<T>::InitialDepositCannotBeZero.into())
+				}
+
+				let single_deposit = compute_deposit_lp_(
+					lp_total_issuance,
+					T::Convert::convert(single.amount),
+					single_balance,
+					single_weight,
+					pool.fee_config.fee_rate,
+				)?;
+
+				T::Assets::transfer(
+					single.asset_id,
+					who,
+					&pool_account,
+					single.amount,
+					keep_alive,
+				)?;
+
+				single_deposit.value
+			},
+			[(first, (first_weight, first_balance)), (second, (second_weight, second_balance))] => {
+				let lp_to_mint = if lp_total_issuance.is_zero() {
+					compute_first_deposit_lp_(
+						&[
+							(T::Convert::convert(first.amount), first_weight),
+							(T::Convert::convert(second.amount), second_weight),
+						],
+						Permill::zero(),
+					)?
+					.value
+				} else {
+					let input_ratio_first_to_second = Permill::from_rational(
+						first.amount,
+						first.amount.safe_add(&second.amount)?,
+					);
+
+					ensure!(
+						per_thing_acceptable_computation_error(
+							input_ratio_first_to_second,
+							first_weight
+						),
+						Error::<T>::IncorrectAmountOfAssets
+					);
+
+					// pass 1 as weight since adding liquidity for all assets
+					// see docs on compute_deposit_lp_ for more information
+					let first_deposit = compute_deposit_lp_(
+						lp_total_issuance,
+						T::Convert::convert(first.amount),
+						first_balance,
+						Permill::one(),
+						pool.fee_config.fee_rate,
+					)?;
+
+					let second_deposit = compute_deposit_lp_(
```

Comment:
NIT: Having this outside of `if_std` make us a pay a runtime cost, which we can avoid.

Reply:
Meaning, we should only compute the second deposit `if_std`?

Reply:
The assertion was supposed to be removed but I seemed to have missed it; if we only need one of the deposits we can just skip calculating the other one(s).

Reply:
No for any constant product pool with any number of assets with any ratio this assertion is always true

Reply:
Well then something's wrong because it was failing in my previous tests - perhaps whatever the issue is/was has been fixed though(?)

In any case, both the assertion & calculation should only be in the if_std block, and the calculation should be nested *in* the debug_assert! to avoid any runtime costs as you said 👍

Perhaps some prop tests for non 50/50 pools for this calculation would be good as well @PoisonPhang?

Reply:
We should defiantly test this before allowing non-50:50 pools (which is not a priority of this release)

However, as it stands, this function is not very testable in this sense. So some other refactoring will be needed.

Reply:
@benluelo 

>Well then something's wrong because it was failing in my previous tests - perhaps whatever the issue is/was has been fixed though(?)

This is formally proven as I sent you before with `appendix c`. This can't go wrong unless there is some rounding related issue, which we should avoid by having an assertion that checks according to an accepted error rate.

>Perhaps some prop tests for non 50/50 pools for this calculation would be good as well

Yes we should definitely try to understand given that you said you had failing tests. Please link them here. I can also check.